lom/src/engine/sequenceHandler.go

```diff
@@ -0,0 +1,830 @@
+ * sequence are complete or timeout, whichever comes earlier.
+ *
+ * When there are multiple sequences ready to go active, they are first sorted
+ * by Priority and then by order of arrival within same priority
+ *
+ * All sequences have timeout. Some sequence may timeout, even before going 
+ * active.
+ *
+ */
+
+type sequenceStatus_t int
+const (
+    sequenceStatus_pending = sequenceStatus_t(iota)
+    sequenceStatus_running
+    sequenceStatus_complete
+)
+
+
+type sequenceState_t struct {
+    sequenceStatus          sequenceStatus_t
+    anomalyInstanceId       string          /* Id referred in all requests & responses for this seq */
+    sequence                *BindingSequence_t /* Read from config; Cache as config may change. */
+    seqStartEpoch           int64           /* Start of the sequence */
+    seqExpEpoch             int64           /* Timepoint of expiry for sequence */
+    ctIndex                 int             /* Index of action in sequence in-progress */
+    context                 []*ActionResponseData /* Ordered responses as received, so far */
+    currentRequest          *activeRequest_t /* Current request in progress */
+    expTimer                *OneShotEntry_t  /* One shot timer fired to watch timeout  */
+}
+
+func (p *sequenceState_t) ExpiryEpoch() int64 {
+    /* If current request expire early, send it; else send seq expiry */
+    if p.currentRequest != nil {
+        if p.currentRequest.reqExpEpoch < p.seqExpEpoch {
+            return p.currentRequest.reqExpEpoch
+        }
+    }
+    return p.seqExpEpoch
+}
+
+func (p *sequenceState_t) validate() bool {
+    if ((len(p.anomalyInstanceId) == 0) ||
+            (p.sequence == nil) ||
+            (p.seqStartEpoch == 0) ||
+            (p.seqExpEpoch < p.seqStartEpoch) ||
+            (p.ctIndex < 1) ||
+            (len(p.context) == 0) ||
+            (p.expTimer == nil)) {
+        LogDebug("sequenceState_t: validate: anId:%v seq:%v epoch(%v/%v) ct(%d) ctx(%v) status(%d) cur(%v) exp(%v)",
+                len(p.anomalyInstanceId) == 0, p.sequence == nil, p.seqStartEpoch,
+                p.seqExpEpoch < p.seqStartEpoch, p.ctIndex, len(p.context) == 0, p.sequenceStatus,
+                p.currentRequest == nil, p.expTimer == nil)
+
+        return false
+    }
+    return true
+}
+
+
+/*
+ * map[anomaly id]sequence_state_t
+ *
+ * Sequences are keyed off of anomaly instance Id 
+ */
+type Sequences_t map[string]*sequenceState_t
+
+/*
+ * timeout:
+ * Two possible candidates:
+ *      1. Currently active request for currently active sequence
+ *      2. All pending sequences, who may timeout before getting chance to execute sequence.
+ */
+type SortedSequences_t []*sequenceState_t
+
+
+type SeqHandler_t struct {
+    /* All Active requests by action name */
+    activeRequests          activeRequestsList_t
+
+    /* Collected sequences by anomaly */
+    sequencesByAnomaly      Sequences_t
```

Comment:
Since a sequence can be accessed by multiple go routines from different methods, should this be made thread safe ?

Reply:
All from single go Routine from engine::runLoop
Hence pretty simple